src/twitter/Timeline.js

```diff
@@ -215,115 +264,141 @@ class Timeline {
-    if (this._timelines.length === 0) {
-      return;
+  is_room_excluded (rooms) {
+    if (!this.config.timelines.poll_if_empty) {
+      const difference = new Set([...rooms].filter(x => !this._empty_rooms.has(x)));
+      if (difference.size === 0) {
+        return true;
+      }
     }
+    return false;
+  }
 
-    var tline = this._timelines[this._t];
-    var req = {
-      user_id: tline.twitter_id,
-      count: TIMELINE_TWEET_FETCH_COUNT,
-      exclude_replies: tline.exclude_replies,
-      tweet_mode: "extended" // https://github.com/Half-Shot/matrix-appservice-twitter/issues/31
-    };
-
-    if(this._newtags.has(req.user_id)) {
-      req.count = 1;
-      this._newtags.delete(req.user_id);
+  is_feed_exceeding_user_limit (tweets, timeline = true) {
+    let max = (timeline ? (TIMELINE_POLL_INTERVAL*this._t) : (HASHTAG_POLL_INTERVAL*this._h))/60000;
+    max = Math.max(max * NEW_PROFILE_THRESHOLD_MIN, 1)
+    if ((timeline ? this.config.timelines : this.config.hashtags).single_account_fallback === true) {
+      const user_ids = new Set(tweets.map((tweet) => {tweet.id_str})).size;
+      log.verbose(`is_feed_exceeding_user_limit: ${user_ids} > ${max}`);
+      return user_ids > max;
     }
+    return false;
+  }
 
-    this.twitter.storage.get_since("@"+tline.twitter_id).then((since) => {
-      log.silly("Polling %s, since value: %s", "@"+tline.twitter_id, since);
-      if (since) {
-        req.since_id = since;
-      }
-      return this.twitter.client_factory.get_client();
-    }).then((client)=>{
-      return client.get('statuses/user_timeline', req).catch((error) =>{
-        log.error("Timeline", "_process_timeline: GET /statuses/user_timeline returned: %s", error.code);
-      });
-    }).then((feed) => {
-      if (feed.length === 0) {
-        return;
-      }
-      else if(feed.length === TIMELINE_TWEET_FETCH_COUNT) {
-        log.info("Timeline poll request hit count limit. Request likely incomplete.");
-      }
-      const s = feed[0].id_str;
-
-      log.silly("Timeline", "Storing since: %s", s);
-      this.twitter.storage.set_since("@"+tline.twitter_id, s);
-
-      // If req.count = 1, the resp will be the initial tweet used to get initial "since"
-      if (req.count !== 1) {
-        this.twitter.processor.process_tweets(tline.room, feed, TWEET_REPLY_MAX_DEPTH);
-      }
-
-      tline.hasProcessedTweets = true;
-    }).catch((err) => {
-      log.error("Timeline", "Error whilst processing timeline %s: %s", tline.twitter_id, err);
-    });
-
+  _process_timeline () {
+    if (this._timelines.length === 0) {
+      return Promise.resolve();
+    }
+    // Rotate to the next timeline.
     this._t++;
     if(this._t >= this._timelines.length) {
       this._t = 0;
     }
+
+    const tline = this._timelines[this._t];
+    return Promise.coroutine(this._process_feed.bind(this))(true, tline);
   }
 
-  _process_hashtags () {
-    if (this._hashtags.length < 1) {
-      return;
+  _process_hashtag () {
```

Comment:
From what I can see, `_process_hashtags` and `_process_timeline` seem to largely do the same thing. It would be nice to factor this stuff out into a single function, which gets given the `timeline` boolean that seems to be what you use elsewhere.